### PR TITLE
LibJS: Implement the resizable ArrayBuffer proposal

### PR DESCRIPTION
**Pull request by PrestonLTaylor**
This pr fully supports resizable ArrayBuffers, as well as support for TypedArray/DataViews to have internal resizable ArrayBuffers.

[Link to proposal](https://tc39.es/proposal-resizablearraybuffer/)

Test262 Diff: ✅ +236 ❌ -236 :^)

### CODE REVIEW

File: Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp [hunk 1]
```diff
@@ -118,21 +118,39 @@ void copy_data_block_bytes(ByteBuffer& to_block, u64 to_index, ByteBuffer& from_
...
+// 25.1.2.1 AllocateArrayBuffer ( constructor, byteLength [ , maxByteLength ] ), https://tc39.es/proposal-resizablearraybuffer/#sec-allocatearraybuffer
+ThrowCompletionOr<ArrayBuffer*> allocate_array_buffer(VM& vm, FunctionObject& constructor, size_t byte_length, Optional<size_t> max_byte_length)
 {
-    // 1. Let obj be ? OrdinaryCreateFromConstructor(constructor, "%ArrayBuffer.prototype%", « [[ArrayBufferData]], [[ArrayBufferByteLength]], [[ArrayBufferDetachKey]] »).
+    // 1. Let slots be « [[ArrayBufferData]], [[ArrayBufferByteLength]], [[ArrayBufferDetachKey]] ».
+
+    // 2. If maxByteLength is present and not empty, then
+    // a. If byteLength > maxByteLength, throw a RangeError exception.
+    // b. Append [[ArrayBufferMaxByteLength]] to slots.
+    if (max_byte_length.has_value() && byte_length > max_byte_length.value())
```
Comment (shannonbooth) [inline Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp:129]: This could be implemented closer to spec text with:
```
    // 2. If maxByteLength is present and not empty, then
   if (max_byte_length.has_value()) {
         // a. If byteLength > maxByteLength, throw a RangeError exception
         if (byte_length > max_byte_length.value()))
            return vm.throw_completion<RangeError>(ErrorType::InvalidByteLengthForResizableArrayBuffer, byte_length, max_byte_length.value());
         // b. Append [[ArrayBufferMaxByteLength]] to slots.
     }
```

Probably same with step c.

File: Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp [hunk 2]
```diff
@@ -118,21 +118,39 @@ void copy_data_block_bytes(ByteBuffer& to_block, u64 to_index, ByteBuffer& from_
...
+    // 6. Set obj.[[ArrayBufferByteLength]] to byteLength.
 
-    // 5. Return obj.
+    // 7. If maxByteLength is present and not empty, then
+    if (max_byte_length.has_value())
+        // a. If it is not possible to create a Data Block block consisting of maxByteLength bytes, throw a RangeError exception.
+        // b. NOTE: Resizable ArrayBuffers are designed to be implementable with in-place growth. Implementations reserve the right to throw if, for example, virtual memory cannot be reserved up front.
+        TRY_OR_THROW_OOM(vm, obj->buffer().try_ensure_capacity(max_byte_length.value()));
+
+    // NOTE: We always want to set this as max_byte_length is represented using an Optional in ArrayBuffer
```
Comment (shannonbooth) [inline Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp:149]: I don't think we need to always set this, the default construct of Optional is empty.

File: Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp [hunk 3]
```diff
@@ -118,21 +118,39 @@ void copy_data_block_bytes(ByteBuffer& to_block, u64 to_index, ByteBuffer& from_
...
     obj->set_buffer(move(block));
 
-    // 4. Set obj.[[ArrayBufferByteLength]] to byteLength.
+    // 6. Set obj.[[ArrayBufferByteLength]] to byteLength.
 
-    // 5. Return obj.
+    // 7. If maxByteLength is present and not empty, then
+    if (max_byte_length.has_value())
+        // a. If it is not possible to create a Data Block block consisting of maxByteLength bytes, throw a RangeError exception.
+        // b. NOTE: Resizable ArrayBuffers are designed to be implementable with in-place growth. Implementations reserve the right to throw if, for example, virtual memory cannot be reserved up front.
```
Comment (shannonbooth) [inline Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp:146]: Multi line if statement should have curlies

File: Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp [hunk 4]
```diff
@@ -201,17 +219,20 @@ ThrowCompletionOr<ArrayBuffer*> array_buffer_copy_and_detach(VM& vm, ArrayBuffer
     if (array_buffer.is_detached())
         return vm.throw_completion<TypeError>(ErrorType::DetachedArrayBuffer);
 
-    // FIXME: 6. If preserveResizability is preserve-resizability and IsResizableArrayBuffer(arrayBuffer) is true, then
+    // 6. If preserveResizability is preserve-resizability and IsResizableArrayBuffer(arrayBuffer) is true, then
     // a. Let newMaxByteLength be arrayBuffer.[[ArrayBufferMaxByteLength]].
     // 7. Else,
     // a. Let newMaxByteLength be empty.
+    auto new_max_byte_length = preserve_resizability == PreserveResizability::PreserveResizability && array_buffer.is_resizable()
```
Comment (shannonbooth) [inline Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp:226]: Can be implemented slightly closer to spec with if/else

File: Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp [hunk 5]
```diff
@@ -201,17 +219,20 @@ ThrowCompletionOr<ArrayBuffer*> array_buffer_copy_and_detach(VM& vm, ArrayBuffer
...
+        ? array_buffer.max_byte_length()
+        : Optional<size_t> {};
 
     // 8. If arrayBuffer.[[ArrayBufferDetachKey]] is not undefined, throw a TypeError exception.
     if (!array_buffer.detach_key().is_undefined())
         return vm.throw_completion<TypeError>(ErrorType::DetachKeyMismatch, array_buffer.detach_key(), js_undefined());
 
     // 9. Let newBuffer be ? AllocateArrayBuffer(%ArrayBuffer%, newByteLength, FIXME: newMaxByteLength).
```
Comment (shannonbooth) [inline Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp:234]: // 9. Let newBuffer be ? AllocateArrayBuffer(%ArrayBuffer%, newByteLength, newMaxByteLength).

**Review by linusg — COMMENTED** [review]
Please add some in-tree tests for this instead of only relying on test262 :)

File: Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp [hunk 6]
```diff
@@ -246,4 +267,62 @@ ThrowCompletionOr<NonnullGCPtr<ArrayBuffer>> allocate_shared_array_buffer(VM& vm
...
 
+// 1.1.3 ArrayBufferByteLength ( arrayBuffer, order ), https://tc39.es/proposal-resizablearraybuffer/#sec-arraybufferlength
+size_t array_buffer_byte_length(VM&, ArrayBuffer& array_buffer, ArrayBuffer::Order)
+{
+    // FIXME: 1. If IsSharedArrayBuffer(arrayBuffer) is true and arrayBuffer has an [[ArrayBufferByteLengthData]] internal slot, then
+    // a. Let bufferByteLengthBlock be arrayBuffer.[[ArrayBufferByteLengthData]].
+    // b. Return ℝ(GetValueFromBuffer(bufferByteLengthBlock, 0, BigUint64, true, order)).
+
+    // 2. Assert: IsDetachedBuffer(arrayBuffer) is false.
+    assert(!array_buffer.is_detached());
```
Comment (shannonbooth) [inline Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp:278]:
```
    VERIFY(!array_buffer.is_detached());
```

We use VERIFY in serenity (also compiled into release)

File: Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp [hunk 7]
```diff
@@ -246,4 +267,62 @@ ThrowCompletionOr<NonnullGCPtr<ArrayBuffer>> allocate_shared_array_buffer(VM& vm
...
+    };
+
+    auto length_storage = LengthStorage { {}, {} };
+
+    // 3. Let getter be a new Abstract Closure with parameters (buffer) that captures lengthStorage and order and performs the following steps when called:
+    auto getter = [length_storage, order](VM& vm, ArrayBuffer& buffer) mutable {
+        // a. If lengthStorage.[[ByteLength]] is empty, then
+        if (!length_storage.byte_length.has_value()) {
+            // i. Assert: lengthStorage.[[ArrayBuffer]] is empty.
+            assert(!length_storage.array_buffer.has_value());
```
Comment (shannonbooth) [inline Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp:302]: VERIFY here also

File: Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp [hunk 8]
```diff
@@ -246,4 +267,62 @@ ThrowCompletionOr<NonnullGCPtr<ArrayBuffer>> allocate_shared_array_buffer(VM& vm
...
+
+            // ii. Set lengthStorage.[[ArrayBuffer]] to buffer.
+            length_storage.array_buffer = { buffer };
+
+            // iii. Set lengthStorage.[[ByteLength]] to ArrayBufferByteLength(buffer, order).
+            length_storage.byte_length = { array_buffer_byte_length(vm, buffer, order) };
+        }
+
+        // b. Assert: SameValue(lengthStorage.[[ArrayBuffer]], buffer) is true.
+        assert(same_value(Value(&length_storage.array_buffer.value()), Value(&buffer)));
```
Comment (shannonbooth) [inline Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp:312]: VERIFY here as well

File: Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp [hunk 9]
```diff
@@ -186,4 +192,72 @@ JS_DEFINE_NATIVE_FUNCTION(ArrayBufferPrototype::transfer_to_fixed_length)
...
+    auto array_buffer_object = TRY(typed_this_value(vm));
+
+    // FIXME: 3. If IsSharedArrayBuffer(O) is true, throw a TypeError exception.
+
+    // 4. If IsDetachedBuffer(O) is true, return +0𝔽.
+    if (array_buffer_object->is_detached())
+        return Value(0);
+
+    // 5. If IsResizableArrayBuffer(O) is true, then
+    // a. Let length be O.[[ArrayBufferMaxByteLength]].
```
Comment (shannonbooth) [inline Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp:209]: Cab be implemented slightly closer to spec here if/else statement

File: Userland/Libraries/LibJS/Runtime/AtomicsObject.cpp [hunk 10]
```diff
@@ -46,36 +46,42 @@ static ThrowCompletionOr<ArrayBuffer*> validate_integer_typed_array(VM& vm, Type
...
-    auto length = typed_array.array_length();
+    // 1. Let getBufferByteLength be MakeIdempotentArrayBufferByteLengthGetter(Unordered).
+    auto get_buffer_byte_length = make_idempotent_array_buffer_byte_length_getter(ArrayBuffer::Order::Unordered);
 
-    // 2. Let accessIndex be ? ToIndex(requestIndex).
+    // 2. Let length be IntegerIndexedObjectLength(typedArray, getBufferByteLength).
+    auto length = integer_indexed_object_length(vm, typed_array, get_buffer_byte_length);
+
+    // 3. Assert: length is not out-of-bounds.
+    assert(length.has_value());
```
Comment (shannonbooth) [inline Userland/Libraries/LibJS/Runtime/AtomicsObject.cpp:59]: VERIFY here

File: Userland/Libraries/LibJS/Runtime/TypedArray.cpp [hunk 11]
```diff
@@ -413,6 +433,141 @@ ThrowCompletionOr<double> compare_typed_array_elements(VM& vm, Value x, Value y,
...
+        return typed_array.array_length().value();
+
+    // 3. Let buffer be O.[[ViewedArrayBuffer]].
+    auto* buffer = typed_array.viewed_array_buffer();
+
+    // 4. Let bufferByteLength be getBufferByteLength(buffer).
+    auto buffer_byte_length = get_buffer_byte_length(vm, *buffer);
+
+    // 5. Assert: IsResizableArrayBuffer(buffer) is true.
+    assert(buffer->is_resizable());
```
Comment (shannonbooth) [inline Userland/Libraries/LibJS/Runtime/TypedArray.cpp:485]: VERIFY

File: Userland/Libraries/LibJS/Runtime/TypedArrayPrototype.cpp [hunk 12]
```diff
@@ -255,101 +254,125 @@ JS_DEFINE_NATIVE_FUNCTION(TypedArrayPrototype::byte_offset_getter)
...
-    auto length = typed_array->array_length();
+    // 3. Let getBufferByteLength be MakeIdempotentArrayBufferByteLengthGetter(SeqCst).
+    auto get_buffer_byte_length = make_idempotent_array_buffer_byte_length_getter(ArrayBuffer::Order::SeqCst);
+
+    // 4. Let len be IntegerIndexedObjectLength(O, getBufferByteLength).
+    auto length = integer_indexed_object_length(vm, *typed_array, get_buffer_byte_length);
 
-    // 4. Let relativeTarget be ? ToIntegerOrInfinity(target).
+    // 5. Assert: len is not out-of-bounds.
+    assert(length.has_value());
```
Comment (shannonbooth) [inline Userland/Libraries/LibJS/Runtime/TypedArrayPrototype.cpp:285]: VERIFY

File: Userland/Libraries/LibJS/Runtime/DataViewPrototype.cpp [hunk 13]
```diff
@@ -154,6 +166,65 @@ static ThrowCompletionOr<Value> set_view_value(VM& vm, Value request_index, Valu
...
+        return view.byte_length().value();
+
+    // 3. Let buffer be view.[[ViewedArrayBuffer]].
+    auto buffer = view.viewed_array_buffer();
+
+    // 4. Let bufferByteLength be getBufferByteLength(buffer).
+    auto buffer_byte_length = get_buffer_byte_length(vm, *buffer);
+
+    // 5. Assert: IsResizableArrayBuffer(buffer) is true.
+    assert(buffer->is_resizable());
```
Comment (shannonbooth) [inline Userland/Libraries/LibJS/Runtime/DataViewPrototype.cpp:187]: VERIFY

**Review by shannonbooth — CHANGES_REQUESTED** [review]
Left a partial review, nice work

File: Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp (hunk 1 above)
Comment (PrestonLTaylor) [reply Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp]: Thanks for the review, i've implemented the reviewed changes :)

**Comment by PrestonLTaylor** [comment]
> Please add some in-tree tests for this instead of only relying on test262 :)

I've added tests into the latest commit :+1:

File: Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp [hunk 14]
```diff
@@ -246,4 +274,62 @@ ThrowCompletionOr<NonnullGCPtr<ArrayBuffer>> allocate_shared_array_buffer(VM& vm
     return obj;
 }
 
+// 1.1.3 ArrayBufferByteLength ( arrayBuffer, order ), https://tc39.es/proposal-resizablearraybuffer/#sec-arraybufferlength
+size_t array_buffer_byte_length(VM&, ArrayBuffer& array_buffer, ArrayBuffer::Order)
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp:278]: `ArrayBuffer const&`

File: Userland/Libraries/LibJS/Runtime/ArrayBuffer.h [hunk 15]
```diff
@@ -45,6 +48,10 @@ class ArrayBuffer : public Object {
         return buffer_impl().size();
     }
 
+    // [[ArrayBufferMaxByteLength]]
+    Optional<size_t> max_byte_length() const { return m_max_byte_length; }
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayBuffer.h:52]: `Optional<size_t> const&`

File: Userland/Libraries/LibJS/Runtime/ArrayBufferConstructor.cpp [hunk 16]
```diff
@@ -64,8 +64,11 @@ ThrowCompletionOr<NonnullGCPtr<Object>> ArrayBufferConstructor::construct(Functi
         return error;
     }
 
-    // 3. Return ? AllocateArrayBuffer(NewTarget, byteLength).
-    return *TRY(allocate_array_buffer(vm, new_target, byte_length_or_error.release_value()));
+    // 3. Let requestedMaxByteLength be ? GetArrayBufferMaxByteLengthOption(options).
+    auto requested_max_byte_length = TRY(get_max_byte_length_option(vm.argument(1)));
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayBufferConstructor.cpp:68]: `auto options = vm.argument(1);` at beginning

File: Userland/Libraries/LibJS/Runtime/ArrayBufferConstructor.cpp [hunk 17]
```diff
@@ -94,4 +97,24 @@ JS_DEFINE_NATIVE_FUNCTION(ArrayBufferConstructor::symbol_species_getter)
     return vm.this_value();
 }
 
+// 1.1.6 GetArrayBufferMaxByteLengthOption ( options ), https://tc39.es/proposal-resizablearraybuffer/#sec-getarraybuffermaxbytelengthoption
+ThrowCompletionOr<Optional<size_t>> ArrayBufferConstructor::get_max_byte_length_option(Value options)
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayBufferConstructor.cpp:101]: Why is this a member function?

File: Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp [hunk 18]
```diff
@@ -186,4 +192,79 @@ JS_DEFINE_NATIVE_FUNCTION(ArrayBufferPrototype::transfer_to_fixed_length)
...
+
+    // NOTE: newByteLength < 0 is always false as new_byte_length is unsigned
+    // 6. If newByteLength < 0 or newByteLength > O.[[ArrayBufferMaxByteLength]], throw a RangeError exception.
+    if (new_byte_length < 0 || new_byte_length > array_buffer_object->max_byte_length().value())
+        return vm.throw_completion<RangeError>(ErrorType::InvalidByteLengthForResizableArrayBuffer, new_byte_length, array_buffer_object->max_byte_length().value());
+
+    // 7. Let hostHandled be ? HostResizeArrayBuffer(O, newByteLength)
+    TRY(host_resize_array_buffer(vm, array_buffer_object, new_byte_length));
+
+    // NOTE: There are more steps in the spec but we can always do the resizing internally, so this step will always return
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp:265]: The very definition of host defined hooks is that they can customize behaviour behind your back, this is not an assumption we can make.

File: Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp [hunk 19]
```diff
@@ -246,4 +274,62 @@ ThrowCompletionOr<NonnullGCPtr<ArrayBuffer>> allocate_shared_array_buffer(VM& vm
...
+        // c. Return lengthStorage.[[ByteLength]].
+        return length_storage.byte_length.value();
+    };
+
+    // 4. Return getter.
+    return getter;
+}
+
+// 1.1.7 HostResizeArrayBuffer ( buffer, newByteLength ) https://tc39.es/proposal-resizablearraybuffer/#sec-hostresizearraybuffer
+ThrowCompletionOr<void> host_resize_array_buffer(VM& vm, ArrayBuffer& array_buffer, size_t new_byte_length)
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp:330]: Check how other `Host*` functions are implemented on the VM, this is not sufficient

File: Userland/Libraries/LibJS/Runtime/AtomicsObject.cpp [hunk 20]
```diff
@@ -217,28 +234,39 @@ static ThrowCompletionOr<Value> atomic_compare_exchange_impl(VM& vm, TypedArrayB
...
-    // 8. Let elementType be TypedArrayElementType(typedArray).
-    // 9. Let elementSize be TypedArrayElementSize(typedArray).
+    // 9. If length is out-of-bounds, throw a TypeError exception.
 
-    // 10. Let isLittleEndian be the value of the [[LittleEndian]] field of the surrounding agent's Agent Record.
+    // 10. If indexedPosition ≥ length, throw a RangeError exception.
+    if (indexed_position >= length)
+        return vm.throw_completion<TypeError>(ErrorType::IndexOutOfRange, indexed_position, length);
+
+    // 11. NOTE: The above checks are not redundant with the checks in ValidateIntegerTypedArray and ValidateAtomicAccess because the call to ToBigInt or ToIntegerOrInfinity on the preceding lines can have arbitrary side effects, which could cause the buffer to become detached. These checks are only needed when
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/AtomicsObject.cpp:252]: only needed when...

File: Userland/Libraries/LibJS/Runtime/AtomicsObject.cpp [hunk 21]
```diff
@@ -46,36 +46,42 @@ static ThrowCompletionOr<ArrayBuffer*> validate_integer_typed_array(VM& vm, Type
     return buffer;
 }
 
-// 25.4.2.2 ValidateAtomicAccess ( typedArray, requestIndex ), https://tc39.es/ecma262/#sec-validateatomicaccess
+// 25.4.2.2 ValidateAtomicAccess ( typedArray, requestIndex ), https://tc39.es/proposal-resizablearraybuffer/#sec-validateatomicaccess
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/AtomicsObject.cpp:49]: For proposals that override an existing methods, keep both spec links and numbers for grepability

```cpp
// 25.4.2.2 ValidateAtomicAccess ( typedArray, requestIndex ), https://tc39.es/ecma262/#sec-validateatomicaccess
// 6.2.2 ValidateAtomicAccess ( typedArray, requestIndex ), https://tc39.es/proposal-resizablearraybuffer/#sec-validateatomicaccess
```

File: Userland/Libraries/LibWeb/Crypto/Crypto.cpp [hunk 22]
```diff
@@ -52,7 +52,7 @@ WebIDL::ExceptionOr<JS::Value> Crypto::get_random_values(JS::Value array) const
     auto& typed_array = static_cast<JS::TypedArrayBase&>(array.as_object());
 
     // 2. If the byteLength of array is greater than 65536, throw a QuotaExceededError and terminate the algorithm.
-    if (typed_array.byte_length() > 65536)
+    if (typed_array.idempotent_byte_length() > 65536)
```
Comment (linusg) [inline Userland/Libraries/LibWeb/Crypto/Crypto.cpp:55]: These changes seem completely unrelated to the commit message (`LibJS: Print the maxByteLength when printing a resizable ArrayBuffer`)

**Review by linusg — CHANGES_REQUESTED** [review]
(partial review)

File: Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp (hunk 19 above)
Comment (PrestonLTaylor) [reply Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp]: Done! host_resize_array_buffer is now defined in the VM in the same way other host functions are.

File: Userland/Libraries/LibWeb/Crypto/Crypto.cpp (hunk 22 above)
Comment (PrestonLTaylor) [reply Userland/Libraries/LibWeb/Crypto/Crypto.cpp:55]: Woops! Rebase shenanigans messed it up :)

**Comment by PrestonLTaylor** [comment]
The proposal is now transitioning into stage 4.

As there seems to be some changes I will close this pr and re-open it when the proposal pr is merged.
